uniffi_bindgen/src/bindings/kotlin/templates/StringHelper.kt

```diff
@@ -38,6 +46,7 @@ public object FfiConverterString: FfiConverter<String, RustBuffer.ByValue> {
     }
 
     override fun write(value: String, buf: ByteBuffer) {
+        check(value)
         val byteArr = value.toByteArray(Charsets.UTF_8)
```

Comment:
Thanks so much for putting this together, @heinrich5991! Let's validate and encode in a single pass, though, like @mhammond suggests.

Since you already have the encoder, you're almost there! I think we can do something like this to make it work:

```suggestion
        val byteArr = Charsets.UTF_8.newEncoder().run {
            onMalformedInput(java.nio.charset.CodingErrorAction.REPORT)
            encode(java.nio.CharBuffer.wrap(value)).array()
        }
```

(Let's also add imports for `java.nio.charset.CodingErrorAction` and `java.nio.CharBuffer` to `wrapper.kt`, so that we don't need to write `java.nio` every time).

Just to note for completeness, we could make this even more efficient by having the encoder write directly into the `RustBuffer`, instead of allocating and copying from `byteArr`. I _definitely_ don't think it's worth doing that now, though. We'd need to know the encoded size to allocate the `RustBuffer`, which means either `maxBytesPerChar()`—overallocating by 4x for almost every string—or `averageBytesPerChar()`, and reallocating if we run out of room.

But validating and encoding in one pass is an easier optimization to make now! 🚀

Reply:
Thanks for the help, I was a little stuck. The provided source fails for some non-ASCII strings, but it's fixed by respecting the returned `ByteBuffer`'s `limit()`.